crates/trie/parallel/src/proof_task.rs

```diff
@@ -173,42 +401,107 @@ where
     }
 
     /// Loops, managing the proof tasks, and sending new tasks to the executor.
+    ///
+    /// # Task Routing
+    ///
+    /// - **Storage Proofs**: Routed to pre-spawned worker pool via bounded channel. Falls back to
+    ///   on-demand spawn if channel is full or disconnected.
+    /// - **Blinded Nodes**: Queued for on-demand execution.
+    ///
+    /// # Shutdown
+    ///
+    /// On termination, `storage_work_tx` is dropped, closing the channel and
+    /// signaling all workers to shut down gracefully.
     pub fn run(mut self) -> ProviderResult<()> {
         loop {
             match self.proof_task_rx.recv() {
-                Ok(message) => match message {
-                    ProofTaskMessage::QueueTask(task) => {
-                        // Track metrics for blinded node requests
-                        #[cfg(feature = "metrics")]
-                        match &task {
+                Ok(message) => {
+                    match message {
+                        ProofTaskMessage::QueueTask(task) => match task {
+                            ProofTaskKind::StorageProof(input, sender) => {
+                                match self
+                                    .storage_work_tx
+                                    .try_send(StorageProofJob { input, result_sender: sender })
+                                {
+                                    Ok(_) => {
+                                        tracing::trace!(
+                                            target: "trie::proof_task",
+                                            "Storage proof dispatched to worker pool"
+                                        );
+                                    }
+                                    Err(crossbeam_channel::TrySendError::Full(job)) => {
+                                        tracing::debug!(
+                                            target: "trie::proof_task",
+                                            "Worker pool queue full, spawning on-demand"
+                                        );
+
+                                        self.on_demand_queue.push_back(
+                                            ProofTaskKind::StorageProof(
+                                                job.input,
+                                                job.result_sender,
+                                            ),
+                                        );
+                                    }
+                                    Err(crossbeam_channel::TrySendError::Disconnected(job)) => {
```

Comment:
Under what circumstances could this occur? If I'm understanding right the worker loop can never error out until its own `work_rx` is disconnected, which would only happen once `ProofTaskMessage::Terminate` is received here. If that's the case then it'd be better to panic or `unreachable!` here, to clarify that this case isn't reachable.